A41-xds-rbac.md

```diff
@@ -0,0 +1,230 @@
+validation failure if present. It is also a validation failure if Permission or
+Principal has a `header` matcher for a `grpc-`-prefixed header name or
+`:scheme`. As described in A39, validation failures for filter configuration
+causes the listener to be NACKed.
+
+The following fields of `Permission` may not be obvious how they map to gRPC:
+
+| Permission Field | gRPC Equivalent |
+| ---------------- | --------------- |
+| header           | Metadata (with caveats below) |
+| url_path         | Fully-qualified RPC method name with leading slash. Same as `:path` header |
+| destination_ip   | Local address for this connection |
+| destination_port | Local port for this connection |
+| metadata         | Hard-coded as empty; never matches |
+| requested_server_name | Hard-coded as empty string |
+
+Because matching supports NOT, the matcher must still be processed even if a
+rule contains references to things that don't generally match; it is not trivial
+to "optimize out" the never-matching rules.
+
+The `header` field is not entirely 1:1 with gRPC Metadata, in part because which
+HTTP headers are present in Metadata is not 100% consistent cross-language.
+For this design, `headers` can include `:method`, `:authority`, and `:path`
+matchers and they should match the values received on-the-wire independent of
+whether they are stored in Metadata or in separate APIs. `:method` can be
+hard-coded to `POST` if unavailable and a code audit confirms the server denies
+requests for all other method types. Implementations must consider the
+request's [hop-by-hop headers][] to not be present. Since hop-by-hop headers
+[are not used in HTTP/2 except for `te: trailers`][rfc7540 connection header],
+transports must consider requests containing the `Connection` header as
+malformed, independent of xDS or RBAC, and only the `TE` header may need special
+handling. If the transport exposes `TE` in Metadata, then RBAC must special-case
+the header to treat it as not present. Multi-valued metadata is represented as
+the concatenation of the values along with a `,` (comma, no added spaces)
+separator, as permitted by HTTP and gRPC. The `Content-Type` provided by the
+client must be used; not a hard-coded value. Binary headers are represented in
+their base64-encoded form, although we rarely expect binary header matchers
+other than presence-checking.
+
+[hop-by-hop headers]: https://datatracker.ietf.org/doc/html/rfc7230#section-6.1
+[rfc7540 connection header]: https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2
+
+As documented for `HeaderMatcher`, Envoy aliases `:authority` and `Host` in its
+header map implementation, so they should be treated equivalent for the RBAC
+matchers; there must be no behavior change depending on which of the two header
+names is used in the _RBAC policy_.
+
+The core gRPC implementation (not just xDS or RBAC) must observe both :authority
+and Host headers. If :authority is missing, Host must be renamed to :authority.
```

Comment:
Isn't case sensitivity defined by the matcher (`StringMatcher.ignore_case`), not the implementation or specific header being matched? I'd agree if you are saying that it is probably a configuration mistake if a matcher for a host was case sensitive. But that doesn't seem part of this specification.

You aren't suggesting that Envoy hard-coded different case matching behavior based on the header name for things like `HeaderMatcher.exact_match`, right? I've honestly not looked much at that sort of detail for HeaderMatcher as we had pre-existing support because of Routes and most things but StringMatcher are deprecated.

Reply:
I don't see any case-insensitivity code in HeaderMatcher for the old-style matchers:
https://github.com/envoyproxy/envoy/blob/327149fdc141ac7e057fb2c37fcf5f2e32047475/source/common/http/header_utility.cc#L152

It really just seems like users should set `StringMatcher.ignore_case=true` for headers like `:authority`.